File: doc-examples/nitrite-example-java/src/main/java/example/TransactionService.java
package example;

import io.micronaut.context.annotation.Primary;
import io.micronaut.transaction.TransactionDefinition;
import io.micronaut.transaction.TransactionOperations;
import io.micronaut.transaction.annotation.Transactional;
import jakarta.inject.Singleton;

/**
 * Demonstrates transactional usage for Nitrite repositories.
 */
@Singleton
@Primary
public class TransactionService {

  private final BookRepository bookRepository;
  private final TransactionOperations<?> transactionOperations;

  public TransactionService(
      BookRepository bookRepository, TransactionOperations<?> transactionOperations) {
    this.bookRepository = bookRepository;
    this.transactionOperations = transactionOperations;
  }

  // tag::transaction-managed[]
  @Transactional
  public void saveBook(String title) {
    bookRepository.save(new Book(title));
  }
  // end::transaction-managed[]

  // tag::transaction-manual-rollback[]
  @Transactional
  public void saveAndRollback(String title) {
    bookRepository.save(new Book(title));
    transactionOperations.executeWrite(status -> {
      status.setRollbackOnly();
      return null;
    });
  }
  // end::transaction-manual-rollback[]

  @Transactional(propagation = TransactionDefinition.Propagation.NOT_SUPPORTED)
  public void logWithoutTransaction(String title) {
    bookRepository.save(new Book(title));
  }
}

File: doc-examples/nitrite-example-kotlin/src/main/kotlin/example/TransactionService.kt
package example

import io.micronaut.transaction.TransactionDefinition
import io.micronaut.transaction.TransactionOperations
import io.micronaut.transaction.annotation.Transactional
import jakarta.inject.Singleton

@Singleton
open class TransactionService(
    private val bookRepository: BookRepository,
    private val transactionOperations: TransactionOperations<*>
) {

  // tag::transaction-managed[]
  @Transactional
  open fun saveBook(title: String) {
    bookRepository.save(Book(title))
  }
  // end::transaction-managed[]

  // tag::transaction-manual-rollback[]
  @Transactional
  open fun saveAndRollback(title: String) {
    bookRepository.save(Book(title))
    transactionOperations.executeWrite { status ->
      status.setRollbackOnly()
      null
    }
  }
  // end::transaction-manual-rollback[]

  @Transactional(propagation = TransactionDefinition.Propagation.NOT_SUPPORTED)
  open fun logWithoutTransaction(title: String) {
    bookRepository.save(Book(title))
  }
}
